### PR TITLE
Greg/gsdx boost texturecache

### PR DESCRIPTION
**Pull request by gregory38**
Greatly reduce the texture removal operation overhead. It might help to improve GSdx speed a little.

### CODE REVIEW

**Comment by refractionpcsx2** [comment]
error C2079: 'GSTextureCache::Source::m_erase_it' uses undefined class 'std::array<std::_List_iterator<std::_List_val<std::_List_simple_types<GSTextureCache::Source *>>>,512>'

**Comment by gregory38** [comment]
:sob: I don't know what is the issue (except the VS suck).

**Comment by refractionpcsx2** [comment]
I'm on it don't worry :P I'll commit it when I fix it

**Comment by refractionpcsx2** [comment]
array<list<Source*>::iterator, MAX_PAGES> m_erase_it;  

saying "incomplete type not allowed"

**Comment by refractionpcsx2** [comment]
also 

"cannot deduce 'auto' type" on line 1989 and 2091 in texturecache

**Comment by gregory38** [comment]
auto will work when m_erase_it is properly detected. What happen if you remove the `::iterator` does it manage to detect the m_erase_it type (obviously it will fail later).

**Comment by refractionpcsx2** [comment]
Nope but if i make it a pointer it works

array<list<Source*>::iterator, MAX_PAGES> *m_erase_it;

**Comment by gregory38** [comment]
Otherwise try to forward declare Source before m_erase_it.
```class Source
blabla m_erase_it
```

**Comment by refractionpcsx2** [comment]
> Otherwise try to forward declare Source before m_erase_it.

What? lol

**Comment by gregory38** [comment]
See example above. Just telling the compiler that Source is a type which might not be yet true inside Source.

**Comment by refractionpcsx2** [comment]
I don't understand your example... what exactly do you want me to do?

**Comment by gregory38** [comment]
http://stackoverflow.com/questions/553682/when-can-i-use-a-forward-declaration

**Comment by turtleli** [comment]
? Not sure how this even compiles on Linux. I think the `<array>` include is missing?

**Comment by gregory38** [comment]
Ah it must be it. I guess it depends on the order of the include (or issue is include from Dx).

**Comment by gregory38** [comment]
let's see if it is better.

**Comment by refractionpcsx2** [comment]
it compiles here now :)

**Comment by gregory38** [comment]
Great. Conclusion, Linux is much better for dev. You give garbage, and it replies with a working program. VS choose the reverse way :stuck_out_tongue: 

Now, let's hope perf is better !

**Comment by refractionpcsx2** [comment]
It's not like Linux not to be anal about something... :P Very out of character!

Going to test a few games now Vs Master

**Comment by gregory38** [comment]
(I hope it isn't slower). The speed impact might be bigger on game that are heavy on texture such as Zone of Ender, baldur gate.

**Comment by refractionpcsx2** [comment]
With MTVU off, Baldurs gate seems possibly 2-5fps faster. so I'm hitting 335fps roughly, instead of 330fps

**Comment by lightningterror** [comment]
2-5 fps is a good improvement.

**Comment by gregory38** [comment]
Faster is still better than slower. Besides it is more robust and might allow to fix some bugs.
It would have been nice to test the Anubis boss on ZoE. Maybe it will be less slow.

**Comment by refractionpcsx2** [comment]
Sorry I don't have ZoE

**Comment by gregory38** [comment]
No worry. Maybe someone will shine.

**Comment by gregory38** [comment]
Could this PR be merged ? So I can continue the work on the GS memory wrapping (#1748)

**Comment by refractionpcsx2** [comment]
It seems to work, so sure :)

File: plugins/GSdx/GSTextureCache.cpp [hunk 1]
```diff
@@ -2070,16 +2074,34 @@ void GSTextureCache::SourceMap::RemoveAt(Source* s)
...
+
+				unsigned long j;
 
-			if(*j == s) {m.erase(j); break;}
+				while(_BitScanForward(&j, p))
+				{
+					// FIXME: this statement could be optimized to a single ASM instruction (instead of 4)
+					// Either BTR (AKA bit test and reset). Depends on the previous instruction.
+					// Or BLSR (AKA Reset Lowest Set Bit). No dependency but require BMI1 (basically a recent CPU)
+					p ^= 1 << j;
```
Comment (turtleli) [inline plugins/GSdx/GSTextureCache.cpp]: I suppose I wonder whether this should actually be `p ^= 1U << j;` (I think 1 << 31 is undefined behaviour, but 1U << 31 is not).

As for merging -  You can go ahead, though I kinda want `std::` to be prefixed onto array and list ;)

**Comment by MarcoEstevez** [comment]
Good job

**Pull request merged**